src/com/zulip/android/HTTPRequest.java

```diff
@@ -1,137 +1,150 @@
-        if (request != null) {
+        if (conn != null) {
             (new AsyncTask<Void, Void, Void>() {
                 @Override
                 protected Void doInBackground(Void... voids) {
-                    request.abort();
+                    conn.disconnect();
                     return null;
                 }
             }).execute();
         }
     }
 
-    // Java doesn't support a request body on DELETE requests, but RFC2616 does
-    // not prohibit it.
-    class HttpDeleteWithReq extends HttpPost {
-        HttpDeleteWithReq(String url) {
-            super(url);
+    private String getPostDataString(HashMap<String, String> params) throws UnsupportedEncodingException {
+        StringBuilder result = new StringBuilder();
+        boolean first = true;
+        for(Map.Entry<String, String> entry : params.entrySet()){
+            if (first)
+                first = false;
+            else
+                result.append("&");
+
+            result.append(URLEncoder.encode(entry.getKey(), "UTF-8"));
+            result.append("=");
+            if(entry.getValue() != null) {
+                result.append(URLEncoder.encode(entry.getValue(), "UTF-8"));
+            }
         }
 
-        @Override
-        public String getMethod() {
-            return "DELETE";
-        }
+        return result.toString();
     }
 
     String execute(String method, String path) throws IOException {
-        AndroidHttpClient httpclient = AndroidHttpClient.newInstance(app
-                .getUserAgent());
-
-        String responseString = null;
+        String responseString = "<empty>";
         try {
-            String url = app.getServerURI() + path;
-            nameValuePairs.add(new BasicNameValuePair("client", "Android"));
-
-            if (method.equals("POST")) {
-                request = new HttpPost(url);
-            } else if (method.equals("PUT")) {
-                request = new HttpPut(url);
-            } else if (method.equals("DELETE")) {
-                request = new HttpDeleteWithReq(url);
-            } else if (method.equals("GET")) {
-                request = new HttpGet(url + "?"
-                        + URLEncodedUtils.format(nameValuePairs, "utf-8"));
+            URL url = new URL(app.getServerURI() + path);
+            this.setProperty("client", "Android");
+            String authstr = this.app.getEmail() + ":" + this.app.getApiKey();
+
+            if (!method.equals("POST") && !method.equals("DELETE") &&
+                    !method.equals("PUT") && !method.equals("GET")){
+                throw new IOException("Wrong HTTP method.");
+            }
+
+            if (method.equals("GET")) {
+                url = new URL(app.getServerURI() + path + "?" + getPostDataString(this.postDataParams));
             }
 
+            conn = (HttpsURLConnection) url.openConnection();
+            conn.setRequestMethod(method);
+            conn.setRequestProperty("Authorization", "Basic "
```

Comment:
Can we switch to [`java.net.Authenticator`][1] instead here?

[1]: https://developer.android.com/reference/java/net/Authenticator.html